### PR TITLE
Somnia bot

### PR DESCRIPTION
**Pull request by Godlander**
testing code for GO discord bot with pando

### CODE REVIEW

File: apps/somnia/src/commands/go.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    const subcommand = interaction.options.getSubcommand();
+    const user = interaction.options.getUser('user', false) || interaction.user;
+    const charname = interaction.options.getString('char', true);
+    let chardata : any;
+    try {
+        chardata = getchardata(user.id, charname);
+    }
+    catch (e) {
+        console.log(e);
+        interaction.reply('Char not found');
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/go.ts:108]: Can you reply with the exception object for now

File: apps/somnia/src/commands/go.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+              {key: e.mainStatKey, value: allStats.art.main[e.rarity][e.mainStatKey][e.level]},
+              ...e.substats.map(s => {
+                if (s.key.endsWith('_')) s.value /= 100;
+                return s;
+              })
+            ]
+          };
+        }))
+      ),
+      selfBuff.common.critMode.add('avg'),
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/go.ts:80]:
```suggestion
      // Enemy
      enemyDebuff.cond.cata.add('spread'),
      enemyDebuff.cond.amp.add(''),
      enemyDebuff.common.lvl.add(80),
      enemyDebuff.common.preRes.add(0.1),
      selfBuff.common.critMode.add('avg'),
```
Probably need the enemy stats, especially the `cata` cond for the dmg to calculate?

File: apps/somnia/src/commands/go.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    ];
+    const calc = genshinCalculatorWithEntries(team);
+    const member0 = convert(selfTag, { member: 'member0', et: 'self' });
+    const read = calc
+      .listFormulas(member0.formula.listing)
+      .find((x) => x.tag.name === 'karma_dmg')!
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/go.ts:86]:
```suggestion
      .find((x) => x.tag.name === 'karma_dmg')
```

File: apps/somnia/src/commands/go.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    return {
+      hp: calc.compute(member0.final.hp).val,
+      atk: calc.compute(member0.final.atk).val,
+      def: calc.compute(member0.final.def).val,
+      eleMas: calc.compute(member0.final.eleMas).val,
+      critRate: calc.compute(member0.final.critRate_).val,
+      critDMG: calc.compute(member0.final.critDMG_).val,
+      karma_dmg: calc.compute(read).val
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/go.ts:94]:
```suggestion
      karma_dmg: read ? calc.compute(read).val : undefined
```

**Review by nguyentvan7 — COMMENTED** [review]
Basic changes for now to hopefully get the dmg calcs working

**Comment by frzyc** [comment]
Looks like the app is hacked into the monorepo. I would recommend generating the app using
```
yarn add -D @nx/node
nx g @nx/node:application somnia-bot --projectNameAndRootFormat derived --framework none
```
and then move the code piece by piece

File: apps/somnia-bot/project.json [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      "outputs": ["{options.outputPath}"],
+      "defaultConfiguration": "production",
+      "options": {
+        "platform": "node",
+        "outputPath": "dist/apps/somnia-bot",
+        "format": ["cjs"],
+        "bundle": false,
+        "main": "apps/somnia-bot/src/main.ts",
+        "tsConfig": "apps/somnia-bot/tsconfig.app.json",
+        "assets": ["apps/somnia-bot/src/assets"],
```
Comment (nguyentvan7) [inline apps/somnia-bot/project.json:18]:
```suggestion
        "assets": [
          "apps/somnia-bot/src/assets",
          {
            "input": "./apps/somnia-bot/src",
            "glob": "config.json",
            "output": "apps/somnia-bot/src/"
          }
        ],
```
Looking at `dist/apps/somnia-bot`, I noticed `config.json` wasn't being copied there. Need to add non-code items as assets here
https://stackoverflow.com/questions/71015454/nx-copy-one-extra-file-to-build-directory
https://nx.dev/recipes/tips-n-tricks/include-assets-in-build

File: .gitignore [hunk 6]
```diff
@@ -55,6 +55,3 @@ Thumbs.db
 
 # Generated JSON files
 *_gen.json
-
-# Discord config file
-apps/somnia/src/config.json
```
Comment (nguyentvan7) [inline .gitignore]: Shouldn't you add the new path here?

Comment (nguyentvan7) [reply .gitignore]: Also, can you upload a blank `config.empty.json` or similar with the needed fields

File: package.json [hunk 7]
```diff
@@ -7,17 +7,20 @@
...
   "devDependencies": {
     "@babel/preset-react": "^7.18.6",
     "@emotion/babel-plugin": "11.11.0",
     "@nx/cypress": "16.4.3",
     "@nx/devkit": "16.4.3",
+    "@nx/esbuild": "16.10.0",
     "@nx/eslint-plugin": "16.4.3",
-    "@nx/jest": "16.4.3",
-    "@nx/js": "16.4.3",
+    "@nx/jest": "16.10.0",
```
Comment (frzyc) [inline package.json:20]: shouldnt have a desync of version of nx package, could cause a lot of problems. would recommend `nx migrate 16.4.3` to sync them up.

File: tsconfig.base.json
[the remark this answers is not in the document]
Comment (frzyc) [reply tsconfig.base.json:1]: You can checkout master version of the file `git checkout master -- tsconfig.base.json`

File: package.json [hunk 8]
```diff
@@ -76,9 +77,11 @@
     "@mui/material": "^5.11.10",
     "@swc/helpers": "0.5.3",
     "core-js": "^3.28.0",
+    "discord.js": "^14.13.0",
     "i18next": "^22.4.10",
     "i18next-browser-languagedetector": "^7.0.1",
     "i18next-http-backend": "^2.1.1",
+    "node-fetch": "^3.3.2",
```
Comment (frzyc) [inline package.json:84]: check if this is still needed

**Pull request closed without merging**